File: src/amrte/web/static/persistence-recovery.js
"use strict";

(() => {
    const API_URL =
        "/api/v1/persistence-recovery";


    function byId(id) {
        return document.getElementById(id);
    }


    function text(value) {
        if (
            value === null ||
            value === undefined ||
            value === ""
        ) {
            return "?";
        }

        return String(value);
    }


    function booleanText(value) {
        return value ? "YES" : "NO";
    }


    function clear(element) {
        while (element.firstChild) {
            element.removeChild(
                element.firstChild
            );
        }
    }


    function detailRow(label, value) {
        const row =
            document.createElement("div");

        row.className = "detail-row";

        const key =
            document.createElement("span");

        key.textContent = label;

        const content =
            document.createElement("strong");

        content.textContent = text(value);

        row.appendChild(key);
        row.appendChild(content);

        return row;
    }


    function renderObject(id, values) {
        const container = byId(id);

        clear(container);

        for (
            const [label, value]
            of Object.entries(values || {})
        ) {
            container.appendChild(
                detailRow(
                    label,
                    value
                )
            );
        }
    }


    function label(name) {
        return String(name)
            .replaceAll("_", " ")
            .replace(
                /\b\w/g,
                character =>
                    character.toUpperCase()
            );
    }


    function renderHistory(history) {
        const body =
            byId("checkpoint-history");

        clear(body);

        const entries =
            Array.isArray(history)
                ? history
                : [];

        byId(
            "history-count"
        ).textContent =
            `${entries.length} GENERATIONS`;

        if (entries.length === 0) {
            const row =
                document.createElement("tr");

            const cell =
                document.createElement("td");

            cell.colSpan = 6;
            cell.textContent =
                "No checkpoint generations.";

            row.appendChild(cell);
            body.appendChild(row);

            return;
        }

        for (const entry of entries) {
            const row =
                document.createElement("tr");

            const values = [
                entry.sequence_number,
                entry.checkpoint_id,
                entry.previous_checkpoint_id,
                entry.recovery_epoch,
                entry.shutdown_status,
                entry.created_at
            ];

            for (const value of values) {
                const cell =
                    document.createElement("td");

                cell.textContent =
                    text(value);

                row.appendChild(cell);
            }

            body.appendChild(row);
        }
    }


    function render(payload) {
        byId(
            "runtime-state"
        ).textContent =
            text(
                payload.runtime.state
            );

        byId(
            "runtime-status-dot"
        ).classList.toggle(
            "running",
            payload.runtime.state
                === "RUNNING"
        );


        const chainValid =
            payload.chain_integrity.valid;

        byId(
            "chain-badge"
        ).textContent =
            chainValid
                ? "CHAIN VALID"
                : "CHAIN INVALID";

        byId(
            "chain-badge"
        ).classList.toggle(
            "success",
            chainValid
        );

        byId(
            "chain-badge"
        ).classList.toggle(
            "warning",
            !chainValid
        );


        byId(
            "metric-checkpoint"
        ).textContent =
            text(
                payload.checkpoint
                    .checkpoint_id
            );

        byId(
            "metric-chain"
        ).textContent =
            chainValid
                ? "VALID"
                : "INVALID";

        byId(
            "metric-epoch"
        ).textContent =
            text(
                payload.runtime
                    .recovery_epoch
            );

        byId(
            "metric-generations"
        ).textContent =
            text(
                payload.storage
                    .checkpoint_count
            );


        renderObject(
            "authority-details",
            {
                "Repository":
                    payload.authority
                        .repository,

                "Source":
                    payload.authority
                        .source,

                "Parallel repository":
                    booleanText(
                        payload.authority
                            .parallel_repository_created
                    ),

                "Mutation enabled":
                    booleanText(
                        payload.authority
                            .mutation_enabled
                    ),

                "Environment":
                    payload.runtime
                        .environment
            }
        );


        renderObject(
            "checkpoint-details",
            {
                "Available":
                    booleanText(
                        payload.checkpoint
                            .available
                    ),

                "Checkpoint ID":
                    payload.checkpoint
                        .checkpoint_id,

                "Sequence":
                    payload.checkpoint
                        .sequence_number,

                "Shutdown status":
                    payload.checkpoint
                        .shutdown_status,

                "Created":
                    payload.checkpoint
                        .created_at,

                "Configuration snapshot":
                    payload.checkpoint
                        .configuration_snapshot_id,

                "Dataset":
                    payload.checkpoint
                        .dataset_id
            }
        );


        renderObject(
            "lineage-details",
            {
                "Checkpoint ID":
                    payload.lineage
                        .checkpoint_id,

                "Previous checkpoint":
                    payload.lineage
                        .previous_checkpoint_id,

                "Sequence":
                    payload.lineage
                        .sequence_number,

                "Generation count":
                    payload.lineage
                        .generation_count,

                "Chain valid":
                    booleanText(
                        payload.chain_integrity
                            .valid
                    )
            }
        );


        renderObject(
            "configuration-details",
            {
                "Minimum interval (seconds)":
                    payload
                        .persistence_configuration
                        .minimum_interval_seconds,

                "Maximum uncheckpointed events":
                    payload
                        .persistence_configuration
                        .maximum_uncheckpointed_events,

                "Retention generations":
                    payload
                        .persistence_configuration
                        .retention_generations,

                "Maximum checkpoint bytes":
                    payload
                        .persistence_configuration
                        .maximum_checkpoint_bytes
            }
        );


        renderObject(
            "recovery-details",
            {
                "Required":
                    booleanText(
                        payload.recovery
                            .required
                    ),

                "Outcome":
                    payload.recovery
                        .outcome,

                "Confidence":
                    payload.recovery
                        .confidence,

                "Ready":
                    booleanText(
                        payload.recovery
                            .ready
                    ),

                "Fallback used":
                    booleanText(
                        payload.recovery
                            .used_fallback
                    ),

                "Recovery epoch":
                    payload.recovery
                        .recovery_epoch,

                "Checkpoint":
                    payload.recovery
                        .checkpoint_id
            }
        );


        renderObject(
            "recovery-readiness",
            {
                "Ready":
                    booleanText(
                        payload
                            .recovery_readiness
                            .ready
                    ),

                "Chain valid":
                    booleanText(
                        payload
                            .recovery_readiness
                            .chain_valid
                    ),

                "Runtime running":
                    booleanText(
                        payload
                            .recovery_readiness
                            .runtime_running
                    ),

                "Recovery report ready":
                    booleanText(
                        payload
                            .recovery_readiness
                            .recovery_report_ready
                    )
            }
        );


        renderObject(
            "storage-details",
            {
                "State root":
                    payload.storage
                        .state_root,

                "Health":
                    payload.storage
                        .health,

                "Checkpoint count":
                    payload.storage
                        .checkpoint_count,

                "Current present":
                    booleanText(
                        payload.storage
                            .current_present
                    ),

                "Previous present":
                    booleanText(
                        payload.storage
                            .previous_present
                    )
            }
        );


        renderObject(
            "execution-boundary",
            {
                "Capability":
                    payload
                        .execution_boundary
                        .capability,

                "Provider":
                    payload
                        .execution_boundary
                        .provider,

                "Financial execution":
                    booleanText(
                        payload
                            .execution_boundary
                            .financial_execution_available
                    )
            }
        );


        const safeguards = {};

        for (
            const [name, value]
            of Object.entries(
                payload
                    .durability_safeguards
                || {}
            )
        ) {
            safeguards[
                label(name)
            ] = (
                typeof value
                === "boolean"
                    ? booleanText(value)
                    : value
            );
        }

        renderObject(
            "durability-safeguards",
            safeguards
        );


        const capabilities = {};

        for (
            const [name, value]
            of Object.entries(
                payload.capabilities
                || {}
            )
        ) {
            capabilities[
                label(name)
            ] = booleanText(value);
        }

        renderObject(
            "capabilities",
            capabilities
        );


        renderHistory(
            payload.checkpoint_history
        );
    }


    function showError(error) {
        const panel =
            byId("error-panel");

        panel.hidden = false;

        byId(
            "error-message"
        ).textContent = (
            error instanceof Error
                ? error.message
                : String(error)
        );

        byId(
            "runtime-state"
        ).textContent = "ERROR";
    }


    async function load() {
        try {
            const response =
                await fetch(
                    API_URL,
                    {
                        method: "GET",
                        headers: {
                            "Accept":
                                "application/json"
                        }
                    }
                );

            if (!response.ok) {
                throw new Error(
                    `HTTP ${response.status}`
                );
            }

            const payload =
                await response.json();

            render(payload);

        } catch (error) {
            showError(error);
        }
    }


    load();
})();

File: src/amrte/web/persistence_recovery.py
from __future__ import annotations

import json
from pathlib import Path
from typing import Any

from amrte.operations.runtime import (
    PersistentResearchRuntime,
)


VERSION = "1.0"


_HISTORY_FIELDS = (
    "checkpoint_id",
    "sequence_number",
    "previous_checkpoint_id",
    "recovery_epoch",
    "created_at",
    "shutdown_status",
    "state_schema_version",
    "application_version",
    "configuration_schema_version",
    "configuration_snapshot_id",
    "configuration_hash",
    "runtime_environment",
    "experiment_id",
    "dataset_id",
    "dataset_fingerprint",
    "payload_hash",
    "manifest_hash",
    "persistence_format_version",
)


def _serialize(value: Any) -> Any:
    name = getattr(
        value,
        "name",
        None,
    )

    if name is not None:
        return name

    return value


def _checkpoint_projection(
    checkpoint: Any,
) -> dict[str, Any]:
    if checkpoint is None:
        return {
            "available": False,
            "checkpoint_id": None,
            "sequence_number": None,
            "previous_checkpoint_id": None,
            "recovery_epoch": None,
            "created_at": None,
            "shutdown_status": None,
            "configuration_snapshot_id": None,
            "configuration_hash": None,
            "dataset_id": None,
            "dataset_fingerprint": None,
        }

    return {
        "available": True,
        "checkpoint_id":
            checkpoint.checkpoint_id,
        "sequence_number":
            checkpoint.sequence_number,
        "previous_checkpoint_id":
            checkpoint.previous_checkpoint_id,
        "recovery_epoch":
            checkpoint.recovery_epoch,
        "created_at":
            checkpoint.created_at.isoformat(),
        "shutdown_status":
            checkpoint.shutdown_status.name,
        "configuration_snapshot_id":
            checkpoint.configuration_snapshot_id,
        "configuration_hash":
            checkpoint.configuration_hash,
        "dataset_id":
            checkpoint.dataset_id,
        "dataset_fingerprint":
            checkpoint.dataset_fingerprint,
    }


def _history_projection(
    repository: Any,
) -> list[dict[str, Any]]:
    history: list[dict[str, Any]] = []

    for path in repository.enumerate_checkpoints():
        try:
            raw = json.loads(
                path.read_text(
                    encoding="utf-8"
                )
            )
        except (
            OSError,
            UnicodeError,
            json.JSONDecodeError,
        ) as exc:
            history.append(
                {
                    "file": path.name,
                    "readable": False,
                    "error":
                        type(exc).__name__,
                }
            )
            continue

        entry = {
            "file": path.name,
            "readable": True,
        }

        for field in _HISTORY_FIELDS:
            entry[field] = raw.get(field)

        history.append(entry)

    history.sort(
        key=lambda item: (
            item.get("sequence_number")
            if isinstance(
                item.get("sequence_number"),
                int,
            )
            else -1
        ),
        reverse=True,
    )

    return history


def build_persistence_recovery_summary(
    runtime: PersistentResearchRuntime,
) -> dict[str, Any]:
    """Build the read-only persistence and recovery view."""

    engine = runtime.engine

    if engine is None:
        raise RuntimeError(
            "AMRTE runtime is not initialized"
        )

    repository = runtime.repository

    if repository is None:
        raise RuntimeError(
            "AMRTE checkpoint repository is unavailable"
        )

    configuration = engine.configuration

    if configuration is None:
        raise RuntimeError(
            "AMRTE configuration is unavailable"
        )

    chain = repository.validate_chain()

    checkpoint = runtime.last_checkpoint

    recovery_report = runtime.recovery_report

    checkpoint_paths = (
        repository.enumerate_checkpoints()
    )

    history = _history_projection(
        repository
    )

    values = configuration.values

    if recovery_report is None:
        recovery = {
            "required": False,
            "outcome": "NEW_RUN",
            "confidence": None,
            "ready": True,
            "used_fallback": False,
            "recovery_epoch":
                runtime.recovery_epoch,
            "checkpoint_id": None,
            "warnings": [],
            "reasons": [],
        }
    else:
        recovery = {
            "required": True,
            "outcome":
                recovery_report.outcome.name,
            "confidence":
                recovery_report.confidence.name,
            "ready":
                recovery_report.ready,
            "used_fallback":
                recovery_report.used_fallback,
            "recovery_epoch":
                recovery_report.recovery_epoch,
            "checkpoint_id": (
                recovery_report.checkpoint.checkpoint_id
                if recovery_report.checkpoint
                is not None
                else None
            ),
            "warnings":
                list(recovery_report.warnings),
            "reasons":
                list(recovery_report.reasons),
        }

    chain_warnings = list(
        getattr(
            chain,
            "warnings",
            (),
        )
    )

    chain_error = None

    if not chain.success:
        error = getattr(
            chain,
            "error",
            None,
        )

        if error is not None:
            chain_error = {
                "code":
                    getattr(
                        error,
                        "code",
                        None,
                    ),
                "message":
                    getattr(
                        error,
                        "message",
                        None,
                    ),
            }

    state_root = getattr(
        repository,
        "root",
        None,
    )

    storage_health = _serialize(
        getattr(
            repository,
            "health",
            None,
        )
    )

    recovery_ready = (
        recovery["ready"]
        and chain.success
    )

    return {
        "version": VERSION,
        "mode": "READ_ONLY",
        "research_only": True,

        "runtime": {
            "state":
                runtime.state,
            "environment":
                engine.environment.name,
            "recovery_epoch":
                runtime.recovery_epoch,
        },

        "authority": {
            "repository":
                type(repository).__name__,
            "source":
                "runtime.repository",
            "parallel_repository_created":
                False,
            "mutation_enabled":
                False,
        },

        "checkpoint":
            _checkpoint_projection(
                checkpoint
            ),

        "lineage": {
            "checkpoint_id": (
                checkpoint.checkpoint_id
                if checkpoint is not None
                else None
            ),
            "previous_checkpoint_id": (
                checkpoint.previous_checkpoint_id
                if checkpoint is not None
                else None
            ),
            "sequence_number": (
                checkpoint.sequence_number
                if checkpoint is not None
                else None
            ),
            "generation_count":
                len(checkpoint_paths),
        },

        "chain_integrity": {
            "valid":
                chain.success,
            "warnings":
                chain_warnings,
            "error":
                chain_error,
        },

        "persistence_configuration": {
            "minimum_interval_seconds":
                values.get(
                    "persistence.minimum_interval_seconds"
                ),
            "maximum_uncheckpointed_events":
                values.get(
                    "persistence.maximum_uncheckpointed_events"
                ),
            "retention_generations":
                values.get(
                    "persistence.retention_generations"
                ),
            "maximum_checkpoint_bytes":
                values.get(
                    "persistence.maximum_checkpoint_bytes"
                ),
        },

        "recovery":
            recovery,

        "storage": {
            "state_root": (
                str(state_root)
                if state_root is not None
                else None
            ),
            "health":
                storage_health,
            "checkpoint_count":
                len(checkpoint_paths),
            "current_present":
                any(
                    path.name
                    == repository.CURRENT
                    for path
                    in checkpoint_paths
                ),
            "previous_present":
                any(
                    path.name
                    == repository.PREVIOUS
                    for path
                    in checkpoint_paths
                ),
        },

        "recovery_readiness": {
            "ready":
                recovery_ready,
            "chain_valid":
                chain.success,
            "runtime_running":
                runtime.state
                == "RUNNING",
            "recovery_report_ready":
                recovery["ready"],
        },

        "checkpoint_history":
            history,

        "durability_safeguards": {
            "minimum_retention_generations":
                2,
            "configured_retention_generations":
                values.get(
                    "persistence.retention_generations"
                ),
            "maximum_checkpoint_bytes":
                values.get(
                    "persistence.maximum_checkpoint_bytes"
                ),
            "chain_verification":
                True,
            "payload_exposure":
                False,
            "automatic_quarantine_from_api":
                False,
        },

        "capabilities": {
            "checkpoint_observation":
                True,
            "chain_verification":
                True,
            "history_observation":
                True,
            "checkpoint_creation":
                False,
            "restore":
                False,
            "rollback":
                False,
            "quarantine":
                False,
            "deletion":
                False,
            "recovery_execution":
                False,
            "configuration_mutation":
                False,
        },

        "execution_boundary": {
            "capability":
                "PROHIBITED",
            "provider":
                type(
                    engine.registry.require(
                        "execution"
                    )
                ).__name__,
            "financial_execution_available":
                False,
        },
    }
